build_tools/buildkite/dana/upload_benchmarks.py

```diff
@@ -0,0 +1,297 @@
+                 verbose=verbose).stdout.strip()
+  segments = info.split(':::')
+  return {
+      'hash': segments[0],
+      'abbrevHash': segments[1],
+      'authorName': segments[2],
+      'authorEmail': segments[3],
+      'subject': segments[4],
+  }
+
+
+def compose_serie_payload(project_id,
+                          serie_id,
+                          serie_description=None,
+                          average_range='5%',
+                          average_min_count=3,
+                          better_criterion='smaller',
+                          override=False):
+  """Composes the payload dictionary for a serie."""
+  payload = {
+      'projectId': project_id,
+      'serieId': serie_id,
+      'analyse': {
+          'benchmark': {
+              'range': average_range,
+              'required': average_min_count,
+              'trend': better_criterion,
+          }
+      },
+      'override': override,
+  }
+  if serie_description is not None:
+    payload['description'] = serie_description
+  return payload
+
+
+def compose_build_payload(project_id,
+                          project_github_comit_url,
+                          build_id,
+                          commit,
+                          override=False):
+  """Composes the payload dictionary for a build."""
+  commit_info = get_git_commit_info(commit)
+  commit_info['url'] = f'{project_github_comit_url}/{commit_info["hash"]}'
+  return {
+      'projectId': project_id,
+      'build': {
+          'buildId': build_id,
+          'infos': commit_info,
+      },
+      'override': override,
+  }
+
+
+def compose_sample_payload(project_id,
+                           serie_id,
+                           build_id,
+                           sample_value,
+                           override=False):
+  """Composes the payload dictionary for a sample."""
+  return {
+      'projectId': project_id,
+      'serieId': serie_id,
+      'sample': {
+          'buildId': build_id,
+          'value': sample_value
+      },
+      'override': override
+  }
+
+
+def get_env_var(var):
+  """Gets the value for an environment variable."""
+  value = os.getenv(var, None)
+  if value is None:
+    raise RuntimeError(f'Missing environment variable "{var}"')
+  return value
+
+
+def post_to_dashboard(url, payload, verbose=False):
+  api_token = get_env_var('IREE_DASHBOARD_API_TOKEN')
+  headers = {
+      'Content-type': 'application/json',
+      'Authorization': f'Bearer {api_token}',
+  }
+  data = json.dumps(payload)
+
+  if verbose:
+    print(f'-api request payload: {data}')
+
+  for attempt in range(RETRY_COUNT):
+    request = requests.post(url, data=data, headers=headers)
+    if verbose:
+      print(f'-api request status code: {request.status_code}')
+    # Return if it succeeded or the content already exists.
+    if request.status_code == 200 or request.status_code == 500:
+      return
+    # Otherwise pause a bit and then retry.
+    time.sleep(1)
```

Comment:
Ooof let's please do exponential backoff if we're going to do retries. I know this isn't a production system, etc. etc. but bad retry strategies are responsible for like half of all outages.

Reply:
Not retrying anymore. We are unlikely to see the server being that busy..